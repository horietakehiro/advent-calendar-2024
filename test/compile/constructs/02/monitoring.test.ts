import {
    aws_cloudwatch_actions as cwActions,
    Duration,
    aws_logs as logs,
    Stack,
} from "aws-cdk-lib";
import {
    LogMetricAlarmFactories,
    LogMetricAlarmFactoryProps,
} from "../../../../lib/constructs/02/monitorings";
import {
    ExtraMatch,
    TypedTemplate,
} from "@horietakehiro/aws-cdk-utul/lib/assertions";
import {
    AWS_CLOUDWATCH_ALARM,
    AWS_LOGS_LOGGROUP,
    AWS_LOGS_METRICFILTER,
} from "@horietakehiro/aws-cdk-utul/lib/types/cfn-resource-types";

describe("LogMetricAlarmFactories", () => {
  const alarmFactories = {
    cseRateAlarm: LogMetricAlarmFactories.cseRateAlarmFactory,
    sseRateAlarm: LogMetricAlarmFactories.sseRateAlarmFactory,
  };
  Object.entries(alarmFactories).forEach(([name, alarmFactory]) => {
    const stack = new Stack();
    const logGroup = new logs.LogGroup(stack, "LogGroup", {});
    const props: LogMetricAlarmFactoryProps = {
      period: Duration.minutes(30),
      threshold: 99,
      actions: [new cwActions.Ec2Action(cwActions.Ec2InstanceAction.REBOOT)],
    };
    alarmFactory(props)(logGroup);
    const template = TypedTemplate.fromStack(stack);
    test(`${name} - 引数で渡したロググループに対してメトリクスフィルターが設定されている`, () => {
      const { id } = template.getResource(AWS_LOGS_LOGGROUP());
      template.hasResource(
        AWS_LOGS_METRICFILTER({
          Properties: {
            LogGroupName: ExtraMatch.ref(id),
          },
        })
      );
    });
    test(`${name} - プロパティで指定した値がアラームの設定に使用されている`, () => {
      const alarm = template.getResource(AWS_CLOUDWATCH_ALARM());
      // 閾値
      expect(alarm.def.Properties?.Threshold).toBe(props.threshold);
      // ピリオド
      const metrics = alarm.def.Properties?.Metrics ?? [];
      expect(metrics.length).toBeGreaterThan(0);
      metrics.forEach((m) => {
        if (m.MetricStat) {
          expect(m.MetricStat.Period).toBe(props.period.toSeconds());
        }
      });
      // アラームアクション
      expect(
        JSON.stringify(alarm.def.Properties?.AlarmActions).includes(
          "ec2:reboot"
        )
      );
    });

    test(`${name} - スナップショットテスト`, () => {
      expect(template.template).toMatchSnapshot(name);
    });

    test("同一のロググループに複数種類のアラームを設定出来る", () => {
      const stack = new Stack();
      const logGroup = new logs.LogGroup(stack, "LogGroup");
      const props: LogMetricAlarmFactoryProps = {
        period: Duration.seconds(60),
        threshold: 99,
        actions: [new cwActions.Ec2Action(cwActions.Ec2InstanceAction.REBOOT)],
      };
      LogMetricAlarmFactories.cseRateAlarmFactory(props)(logGroup);
      LogMetricAlarmFactories.sseRateAlarmFactory(props)(logGroup);

      const template = TypedTemplate.fromStack(stack);
      const lg = template.getResource(AWS_LOGS_LOGGROUP());
      template.allResourcesProperties(
        AWS_LOGS_METRICFILTER({
          Properties: {
            LogGroupName: ExtraMatch.ref(lg.id),
          },
        })
      );
    });
  });
});
