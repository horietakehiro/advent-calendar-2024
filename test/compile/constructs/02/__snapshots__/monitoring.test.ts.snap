// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogMetricAlarmFactories cseRateAlarm - スナップショットテスト: cseRateAlarm 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogGroupAllRequestCountE37893C0": {
      "Properties": {
        "FilterPattern": "[..., agent != "ELB-HealthChecker*", xForwardedFor]",
        "LogGroupName": {
          "Ref": "LogGroupF5B46931",
        },
        "MetricTransformations": [
          {
            "DefaultValue": 0,
            "MetricName": "AllRequestCount",
            "MetricNamespace": "WebApp",
            "MetricValue": "1",
            "Unit": "Count",
          },
        ],
      },
      "Type": "AWS::Logs::MetricFilter",
    },
    "LogGroupCSECountDBFD5086": {
      "Properties": {
        "FilterPattern": "[..., status_code = 4*, bytes, referer, agent != "ELB-HealthChecker*", xForwardedFor]",
        "LogGroupName": {
          "Ref": "LogGroupF5B46931",
        },
        "MetricTransformations": [
          {
            "DefaultValue": 0,
            "MetricName": "CSECount",
            "MetricNamespace": "WebApp",
            "MetricValue": "1",
            "Unit": "Count",
          },
        ],
      },
      "Type": "AWS::Logs::MetricFilter",
    },
    "LogGroupF5B46931": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "cseRateAlrm30DC54C5": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":automate:",
                {
                  "Ref": "AWS::Region",
                },
                ":ec2:reboot",
              ],
            ],
          },
        ],
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "(cse / (all+0.01)) * 100",
            "Id": "expr_1",
          },
          {
            "Id": "cse",
            "MetricStat": {
              "Metric": {
                "MetricName": "CSECount",
                "Namespace": "WebApp",
              },
              "Period": 1800,
              "Stat": "Sum",
              "Unit": "Count",
            },
            "ReturnData": false,
          },
          {
            "Id": "all",
            "MetricStat": {
              "Metric": {
                "MetricName": "AllRequestCount",
                "Namespace": "WebApp",
              },
              "Period": 1800,
              "Stat": "Sum",
              "Unit": "Count",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 99,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`LogMetricAlarmFactories sseRateAlarm - スナップショットテスト: sseRateAlarm 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogGroupF5B46931": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "LogGroupSSECountF5A20371": {
      "Properties": {
        "FilterPattern": "[..., request, status_code = 5*, bytes, referer, agent != "ELB-HealthChecker*", xForwardedFor]",
        "LogGroupName": {
          "Ref": "LogGroupF5B46931",
        },
        "MetricTransformations": [
          {
            "DefaultValue": 0,
            "MetricName": "SSECount",
            "MetricNamespace": "WebApp",
            "MetricValue": "1",
            "Unit": "Count",
          },
        ],
      },
      "Type": "AWS::Logs::MetricFilter",
    },
    "SSERateAlrm87C530C7": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":automate:",
                {
                  "Ref": "AWS::Region",
                },
                ":ec2:reboot",
              ],
            ],
          },
        ],
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "(sse / (all+0.01)) * 100",
            "Id": "expr_1",
          },
          {
            "Id": "sse",
            "MetricStat": {
              "Metric": {
                "MetricName": "SSECount",
                "Namespace": "WebApp",
              },
              "Period": 1800,
              "Stat": "Sum",
              "Unit": "Count",
            },
            "ReturnData": false,
          },
          {
            "Id": "all",
            "MetricStat": {
              "Metric": {
                "MetricName": "AllRequestCount",
                "Namespace": "WebApp",
              },
              "Period": 1800,
              "Stat": "Sum",
              "Unit": "Count",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 99,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
